src/amount.rs

```diff
@@ -0,0 +1,687 @@
+///
+/// Likewise, if/when the currency gains in real world value, a smaller
+/// amount of tokens can buy more.  As such, it can then make sense to
+/// specify amounts with a smaller (negative exponent) unit.  This can happen
+/// organically, as with the USD dollar where people used to buy eggs with
+/// a nickel or dime, but now buy them with dollars.  The market has moved
+/// to a higher unit simply because that requires the least number of coins/tokens.
+///
+/// For a deflationary currency gaining in value such as our DBC's are expected to be,
+/// the market movement should normally be towards smaller units rather than larger.
+///
+/// Any two Amount can be added and subtracted only if they can both be represented
+/// in the same unit, without the operation under or overflowing counter.
+///
+/// In other words, it does not make sense to try and add eg:
+///     Amount{count: 15, unit: 0}     (aka 15)  and
+///     Amount{count: 1, unit: -20}    (aka 0.00000000000000000001)
+///
+/// If we normalize both of these to unit: -20, then we have:
+///     Amount{count: 1500000000000000000000, unit: -20}     (aka 15)  and
+///     Amount{count: 1, unit: -20}    (aka 0.00000000000000000001)
+///
+/// However 1500000000000000000000 overflows our counter, which only
+/// allows values up to 1 billion.  Hence these two amounts are incompatible.
+///
+/// Since the amounts cannot even add or subtract if they are not close
+/// enough together, the Mint will not be able to sum inputs or outputs
+/// that are too far apart and will issue an error.
+///
+/// This prevents users/wallets from generating huge amounts
+/// of change with very unlike denominations, eg by subtracting 1 10^-30 from 1 10^3
+/// This is a problem when using eg u128 to represent Amounts.  In the worst case with
+/// u128 approx 40 outputs can be created when subtracting 1u128 from u128::MAX.
+///
+/// By comparison, using this Amount scheme with 1 billion max count, the max
+/// change outputs is 9.
+///
+/// Unfortunately when using random number generators for quicktest test cases,
+/// the common case becomes near the worst case.  Also, large numbers of inputs and
+/// outputs create load on our system, in particular for signing and verifying.
+/// Thus we are incentivized to keep the number of change coins as low as we
+/// reasonably can.
+///
+/// In effect, this Amount scheme makes it hard for users to generate essentially
+/// worthless dust amounts by accident. It is possible to do if one really tries
+/// by reissuing in ever smaller amounts, but wallet software should generally
+/// be trying NOT to do that.  And if a user does manage it, then s/he will have
+/// difficulty using them in transactions with other people.  Fortunately the
+/// reverse process can be used to bring them up into a "normal" range again.
+///
+#[derive(Clone, Debug, Copy, Default, Serialize, Deserialize)]
+pub struct Amount {
+    pub count: AmountCounter,
+    pub unit: PowerOfTen,
+}
```

Comment:
the range of values this `Amount` struct can take is `domain(AmountCounter) * domain(PowerOfTen)` which gets very large if the range of `AmountCounter` is 1billion, with a single power of ten we have 1billion anonymity classes right? That would seem like enough to break traceability.

Can we restrict this count down closer to 0..9 to get 10 anonymity classes per magnitude?

Reply:
yeah, so the Amount struct is representing transactable amounts, such as prices on an item.   Using USD as an example, if you sell me your old laptop and you ask $357 for it, well there is no $357 bill.  So instead I give you three `100` one `50` one `5` and two `one`.    Taking this back to our DBC denom scheme, I would give you one `300`, one `50` and one `7`.  There is always just one coin/bill per digit.

So the anonymity pools being used are those of the denominations:  a pool of `300`, another of `50` and another of `7`.   The pool for each is the set of all transactions that use the same denomination.

Now one can say that the transacted amount itself, 357 is a unique and interesting number.   And it is.  However it is unique to that single transaction.  So an observer can see that $357 was transacted there, but cannot link to any other tx by 357.  Only by the 300, 50, and 7, which are in very common (constant) usage by others.

Reply:
We can certainly lower the counter_max() number.  Doing so would reduce outliers, eg the number of people transacting 1 billion USD is going to be pretty small, so those tx could stand out.  However, it also forces people to use less granular amounts in their transactions.  This might annoy people, so I don't think we should get too carried away there.  Maybe we could consider going as low as 10^6.   Anyway, it's a knob we can twist.... lowering it both increases privacy (few outliers) and reduces work for the mint.